client/analytics/report/index.js

```diff
@@ -34,51 +35,48 @@ const getReportParam = ( { params, path } ) => {
 	return params.report || path.replace( /^\/+/, '' );
 };
 
-class Report extends Component {
-	constructor() {
-		super( ...arguments );
-
-		this.state = {
-			hasError: false,
-		};
-	}
-
-	componentDidCatch( error ) {
-		this.setState( {
-			hasError: true,
-		} );
+const ErrorLogger = ( { error } ) => {
+	useEffect( () => {
 		/* eslint-disable no-console */
 		console.warn( error );
```

Comment:
I see a few issues with this PR, it feels like they're all a bit inter-related so I'll try summarise:

1. A nitpick but we don't really have an approach to logging in production right now, so I'd prefer we don't until we've thought about that in depth

2. This PR is 2 things with separate concerns. 1 is adjusting the error handling of the Report component and warrants its own discussion, the other is a new error boundary component, which in and of itself is probably less controversial and easy to reason about.

3. I also think we're missing tests on both subjects. the error boundary should have at least a basic test of functionality to cover against regressions and the new report behaviour definitely needs some tests.

4. Without tests its hard for me to reason exactly about the report changes you've made here. For example I notice in your code that you don't display the `ReportUi` if the error boundary caught an error, but how does that interact with `isError` currently? 

5. It would have been great if you could have talked about what situation was causing a fatal here so that we could understand the problem and why you introduced it to this component.

6. One big question for me is: do we need an ErrorBoundary component in packages? It doesn't provide anything design wise, and functionality wise its basically a copy/paste of the example from the React docs, so what value does increasing the size of `@woocommerce/components` have with this component? I'd argue it should at least start as private to this code base. I'd prefer to see a component that provides a piece of designed error UI for fatals but we'd need some input from design on that. I think it'd be valuable though.

All of this is to say that I think the reasoning for this PR is very unclear to me and in its current state I don't think these changes make a lot of sense, but a big part of my misunderstanding here could be that there is not a detailed rationale provided in the description.

Reply:
Hey @samueljseay , thanks for taking a look!

> A nitpick but we don't really have an approach to logging in production right now, so I'd prefer we don't until we've thought about that in depth

That's fair! However, I'd like to keep the `ErrorBoundary` component unaware of logging behavior, since my idea is that it would be used by third-party consumers.

> This PR is 2 things with separate concerns. 1 is adjusting the error handling of the Report component and warrants its own discussion, the other is a new error boundary component, which in and of itself is probably less controversial and easy to reason about.

I initially thought about just introducing the `ErrorBoundary` component. But then I noticed that there's another example in the codebase that uses `componentDidCatch`.
I know that when PRs are reviewed there can always be some discussion on "there's some existing code that uses X, do you think it can be refactored?" or "when do you think we can refactor the existing code to use this new approach?". By keeping that in mind, I wanted to limit the discussion on that topic by just refactoring everything - but it looks like it backfired on me :D 
Let me know, if you'd like I can revert the changes on the Report component.

> I also think we're missing tests on both subjects. the error boundary should have at least a basic test of functionality to cover against regressions and the new report behaviour definitely needs some tests.

I added tests for the ErrorBoundary component on `packages/components/src/error-boundary/test/index.js` - do you have specific test cases in mind that I'm missing?
I tried to keep the behavior of the `Report` component unchanged. In case of error, it still doesn't display its children and logs the error in the console. Otherwise the UI should still behave the same.

> Without tests its hard for me to reason exactly about the report changes you've made here. For example I notice in your code that you don't display the ReportUi if the error boundary caught an error, but how does that interact with isError currently?

It also currently does not display the UI of the Report component, in the case the children of `Report` throw an error: https://github.com/woocommerce/woocommerce-admin/blob/main/client/analytics/report/index.js#L56-L58
It looks like `isError` prop comes from a selector and is separate from the error state: https://github.com/woocommerce/woocommerce-admin/blob/main/client/analytics/report/index.js#L108
To be honest, I didn't investigate much on this, I just made sure that with my refactor the code would behave the same.

> It would have been great if you could have talked about what situation was causing a fatal here so that we could understand the problem and why you introduced it to this component.
> One big question for me is: do we need an ErrorBoundary component in packages? It doesn't provide anything design wise, and functionality wise its basically a copy/paste of the example from the React docs, so what value does increasing the size of @woocommerce/components have with this component? I'd argue it should at least start as private to this code base. I'd prefer to see a component that provides a piece of designed error UI for fatals but we'd need some input from design on that. I think it'd be valuable though.

The main reason I want to introduce this component is for utility on consumers of `@woocommerce/components`. It looks like there's a few "utility" components in there too, so I thought it would be a good place for `ErrorBoundary` as well.
One of the places I thought it would be useful (which is in my case) is for consumers of the `woocommerce_admin_onboarding_task_list` filter. If the [`container` attribute](https://github.com/woocommerce/woocommerce-admin/blob/main/docs/examples/extensions/add-task/js/index.js#L85) throw some error, the user is presented with a white screen of death. Consumers could wrap their tree with `ErrorBoundary` and more easily show an error message: `container: <ErrorBoundary> { ( { error } ) => error ? <ErrorUi /> : <Task /> }</ErrorBoundary>`.
The error could occur when an exception is thrown when interacting with UI elements or when fetching data in a `useEffect` hook, for example.
This is one case that I wanted to cover in `woocommerce-payments`: https://github.com/Automattic/woocommerce-payments/pull/1734/files#diff-4a193dcb5f66a0b4af8dba93a5049e6dc7095a7874e80a3073a9da0516bb5f5aR28
I would argue to also use it within other components as well.
There currently is no way to use the `componentDidCatch` lifecycle method unless you're using a class-based component. By encapsulating the `componentDidCatch` into its own component, consumers can just write functional components without having to rely on "older" class components and having to refactor the code later.
I would avoid having any sort of UI within `ErrorBoundary`, to keep things simple. I just want this component to be as bare-bone as possible, to allow consumers to implement their own UI. We can always think about having _another_ component that maybe is colocated with `ErrorBoundary`, whose responsibility is to just display the error message. Maybe an `ErrorUI` that could be used like `<ErrorBoundary>{ ( { error } ) => error ? <ErrorUI error={ error } /> : <UI /> }</ErrorBoundary>`.

I hope I clarified and touched on all your points! :) Let me know your thoughts and/or if you'd like me to revert the changes on the `Report` component!

Reply:
Instead of starting a new thread, I will respond here given it is related to this thread. @frosso thanks for the detailed response to @samueljseay thoughts. I will try my best to cover most points brought up by the both of you, and responded to the more specific points down below.

I am leaning towards Sam's sentiment here: ` It doesn't provide anything design wise, and functionality wise its basically a copy/paste of the example from the React docs`. Not saying, that we aren't in need of a ErrorBoundary, because I think we are. But, I think in its current state it's easier for 3PD's to create their own ErrorBoundary with their own UI.

I do see a case where we add an ErrorBoundary component to `@woocommerce/components`, but one that includes a visual error screen that reflects the WooCommerce design. Allowing 3PD's to keep the same look and feel for error handling. 
We do have an open issue to add this #3406 and it might be time to shine some new light on that ticket and come up with a standardized design.

My proposition if it suites the both of you (I am open to other suggestions, and if this has to increase in priority)?
- Revert these changes for now, and have another look at #3406, seeing if we could get a standardized WooCommerce error screen.
- Leave it up to the 3PD's to handle their own errors (for now).
- **If** there is a specific error happening in the `Reports` component that needs to be caught, then move the `ErrorBoundary` to `client` and make use of it in `Reports`. If not, we should wait until we solidified #3406.  

> By keeping that in mind, I wanted to limit the discussion on that topic by just refactoring everything - but it looks like it backfired on me :D

🤣 it happens to the best of us 😄 

> Let me know, if you'd like I can revert the changes on the Report component.

As mentioned above, I think we should revert the changes in Report component, unless there is a specific error that needs to be caught (that you know off).

> I added tests for the ErrorBoundary component on packages/components/src/error-boundary/test/index.js - do you have specific test cases in mind that I'm missing?

Thanks for adding the tests 😄 

> The main reason I want to introduce this component is for utility on consumers of @woocommerce/components. It looks like there's a few "utility" components in there too, so I thought it would be a good place for ErrorBoundary as well.

Totally, and I think it would be worth it if it comes with a WooCommerce design (that is a little bit customizable) as mentioned above.

> One of the places I thought it would be useful (which is in my case) is for consumers of the woocommerce_admin_onboarding_task_list filter. If the container attribute throw some error, the user is presented with a white screen of death.

I think this is a very valid point, and we should probably add a boundary for this sooner then later, this might be one of those use cases where it should re-direct back to home screen and just show an error notice (on our end).
I believe it should be up to the consumers to handle errors on their end however they see fit.

> I hope I clarified and touched on all your points! :) Let me know your thoughts and/or if you'd like me to revert the changes on the Report component!

Yeah the detailed response is very much appreciated. I hope you understand the direction @samueljseay and I are coming from.

Reply:
@louwie17 thanks for the additional information!
I think all the comments are completely fair 👍  I'm going to close this PR in favor of a more complete https://github.com/woocommerce/woocommerce-admin/issues/3406 , it's all good :)